src/MongoDB.Driver/TransactionExecutor.cs

```diff
@@ -0,0 +1,284 @@
+                    {
+                        return new CallbackOutcome<TResult>(default, true);
+                    }
+
+                    if (mongoException.HasErrorLabel(UnknownTransactionCommitResultLabel))
+                    {
+                        await clientSession.CommitTransactionAsync(cancellationToken).ConfigureAwait(false);
+                    }
+                }
+
+                throw;
+            }
+        }
+
+        private static bool CommitWithRetries(IClientSessionHandle clientSession, DateTime startTime, IClock clock, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    clientSession.CommitTransaction(cancellationToken);
+                    return true;
+                }
+                catch (Exception ex)
+                {
+                    if (ex is MongoException mongoException)
+                    {
+                        var now = clock.UtcNow;
+                        if (mongoException.HasErrorLabel(UnknownTransactionCommitResultLabel) &&
+                            !HasTimedOut(startTime, now))
+                        {
+                            continue;
+                        }
+
+                        if (mongoException.HasErrorLabel(TransientTransactionErrorLabel) && !HasTimedOut(startTime, now))
+                        {
+                            // the transaction will be full processed once again
+                            return false;
+                        }
+                    }
+
+                    throw;
+                }
+            }
+        }
+
+        private static async Task<bool> CommitWithRetriesAsync(IClientSessionHandle clientSession, DateTime startTime, IClock clock, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+
+                    await clientSession.CommitTransactionAsync(cancellationToken).ConfigureAwait(false);
+                    return true;
+                }
+                catch (Exception ex)
+                {
+                    if (ex is MongoException mongoException)
+                    {
+                        var now = clock.UtcNow;
+                        if (mongoException.HasErrorLabel(UnknownTransactionCommitResultLabel) &&
+                            !HasTimedOut(startTime, now))
+                        {
+                            continue;
+                        }
+
+                        if (mongoException.HasErrorLabel(TransientTransactionErrorLabel) && !HasTimedOut(startTime, now))
+                        {
+                            // the transaction will be full processed once again
+                            return false;
+                        }
+                    }
```

Comment:
Done
Note: I've added:
    
    var now = clock.UtcNow

for tests reasons, because of `mock` behavior. I set different time values for each next `clock` calling. So, in a catch block we need to have only one expected time value, but in proposed code, mock logic will set a different time to each `clock` calling.